plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2741,7 +2741,7 @@ void producer_plugin_impl::produce_block() {
    br.total_time += fc::time_point::now() - start;
    chain.commit_block(br);
 
-   const auto& new_b = chain.head().block();
+   const signed_block_ptr new_b = chain.head().block();
```

Comment:
In this particular case, I wonder if it would be better to replace the usages of `new_b` with inline calls to `chain.head().xxx`.
Also why does `head()` return a `block_handle` rather than a `const block_handle&`. Seems like a pointless copy of `block_handle` on every access.

Reply:
> Also why does `head()` return a `block_handle` rather than a `const block_handle&`

`fork_db_head()` does construct a handle, so returning a value is really the only thing we can do I tbelieve unless we add a new `block_handle` member in fork_db.

As for `head()`, I think we should return a reference, with the hope that users in tests will not store the returned reference in an `auto&`  variable which would then change along with controller's `chain_head`.

Reply:
I think we should be consistent in what `head()` and `fork_db_head()` returns. I'm tempted to suggest we add a `const block_handle& chain_head() const` to avoid the `std::variant` of `std::shared_ptr` copies. However, I don't think it is worth it. Maybe we should instead have `block_handle::block()` return a `signed_block_ptr`.

Reply:
> Maybe we should instead have `block_handle::block()` return a `signed_block_ptr`.

+1.

 I think also `block_handle::id()` and `previous()` should return values not references.

Reply:
What do you all want to do about this? Go ahead as is or go through some more refactor?

Reply:
At this point, probably just merge this PR. We can refactor later if we think it is needed.